[Draft] Save values in memory

m-mora opened the pull request:
This is the initial code, still not use it
it is for review and clean up

m-mora committed:
Initial version to save values in memory

Signed-off-by: Marco Mora <m.mora@intel.com>

m-mora committed:
move implementation of functions to cpp file

m-mora commented on src/system/storage/storage.h:
```diff
@@ -0,0 +1,64 @@
+{
+private:
+  int num_releys;
+  conf_time_t reley;
```
maybe change reley -> device

Elemeants replied:
What means device?
Shouldn't be more like `RelayConfigTime`

Elemeants replied:
Also I don't understand why are u using only one struct, shouldn't be an array of configurations?

m-mora replied:
fair with the name
on regards of using just one struct is just to have granularity, can manage each configuration separately

Elemeants commented on src/system/storage/storage.h:
```diff
@@ -0,0 +1,64 @@
+#include <EEPROM.h>
+#include "../../utils/logger.h"
+
+#define BLYNK_TEMPLATE_ID_MAX     20  // Max number of characters
+#define BLYNK_TEMPLATE_NAME_MAX   20
+#define BLYNK_AUTH_TOKEN_MAX      40
+#define BLYNK_TEMPLATE_ID_BASE    0   // base address in bytes
+#define BLYNK_TEMPLATE_NAME_BASE  20
+#define BLYNK_AUTH_TOKEN_BASE     40
```
Shouldn't be this defined on the WifiManager or on a Blynk specific header?

m-mora replied:
The idea here is to have the ability to save credentials, maybe remove "BLYNK" from the name, but it can be used for other platform that most likely will require kind of the same data.
To have something like
```
0  |------------|
   |            |  <-- space to save credentials
   |            |
80 |------------|
   |            |  <-- space to save reley 1 conf
   |------------|
   |            |  <-- space to save reley 2 conf
   |------------|
   |            |
   |------------|
   |            |
   |------------|
```

Elemeants commented on src/system/storage/storage.h:
```diff
@@ -0,0 +1,64 @@
+  uint8_t sec;
+  uint8_t duration;
+  uint8_t days;
+} conf_time_t;
```
I think we need to be more specific what is the purpose of this struct.
```suggestion
} eeprom_map_config_time_t;

```

m-mora replied:
fair

Elemeants commented on src/system/storage/storage.h:
```diff
@@ -0,0 +1,64 @@
+String templateName;
+String authToken;
+
+class storage
```
We use PascalCase, so this class should follow that.
```suggestion
class Storage
```

Elemeants commented on src/system/storage/storage.h:
```diff
@@ -0,0 +1,64 @@
+  conf_time_t reley;
+  
+public:
+  storage(int _num_releys);
```
I think the number of relays should be defined on the eeprom also.

So we have a general struct that represents our data header and then a buffer with the configuration entries for each relay.

m-mora replied:
I commented this before

m-mora committed:
Changed defines to structimplemented some feedback

m-mora committed:
implemented some feedback

Elemeants commented on src/system/storage/storage.h:
```diff
@@ -0,0 +1,64 @@
+  uint8_t min;
+  uint8_t sec;
+  uint8_t duration;
+  uint8_t days;
```
Add 11  bytes extra for future definitions.

Elemeants commented on src/system/storage/storage.h:
```diff
@@ -0,0 +1,64 @@
+#include <EEPROM.h>
+#include "../../utils/logger.h"
+
+#define TEMPLATE_ID_MAX     20  // Max number of characters
+#define TEMPLATE_NAME_MAX   20
+#define AUTH_TOKEN_MAX      40
+#define TEMPLATE_ID_BASE    0   // base address in bytes
+#define TEMPLATE_NAME_BASE  20
+#define AUTH_TOKEN_BASE     40
+
+#define CONF_MEM_START  80            // leave the first space for credentials
```
```suggestion
typedef struct {
  uint8_t templateId [20];
  uint8_t templateName [20];
  uint8_t authToken [20];
} EEPROM_CredentialStorage_t;

#define CONF_MEM_START sizeof(EEPROM_CredentialStorage_t)
```

m-mora committed:
Merge branch 'dev/mmora/storage' of ssh://github.com/klicGdl/IrrigationSystem into dev/mmora/storage

m-mora committed:
Implemented feedback

Elemeants commented on src/system/storage/storage.cpp:
```diff
@@ -0,0 +1,79 @@
+
+bool Storage::getCredentials(String templateID, String templateName, String authToken)
+{
+  EEPROM.get(offsetof(EEPROM_CredentialStorage_t, templateid),templateID);
```
Where do you specify that it is an string or its size?

m-mora replied:
it doesn't, this function works as a sequence of bytes
I did some tests ouside and it can recover the tokens and IDs

Elemeants commented on src/system/storage/storage.cpp:
```diff
@@ -0,0 +1,79 @@
+  hour    = ReleyConfigTime.hour;
+  minute  = ReleyConfigTime.min;
+  second  = ReleyConfigTime.sec;
+  duration  = ReleyConfigTime.duration;
```
Align, please.

m-mora replied:
Sue

m-mora committed:
Merge branch 'dev/mmora/storage' of ssh://github.com/klicGdl/IrrigationSystem into dev/mmora/storage

Elemeants commented on src/system/storage/storage.cpp:
```diff
@@ -0,0 +1,79 @@
+
+bool Storage::getConfiguration(int releyID, uint8_t hour, uint8_t minute, uint8_t second, uint8_t duration, uint8_t days)
+{
+  EEPROM.get(CONF_MEM_START * releyID, ReleyConfigTime);
```
Apply the Credential offset
```suggestion
  EEPROM.get((sizeof(eeprom_map_conf_time_t) * releyID) + CONF_MEM_START, ReleyConfigTime);
```

Elemeants commented on src/system/storage/storage.cpp:
```diff
@@ -0,0 +1,79 @@
+  return true;
+}
+
+bool Storage::getConfiguration(int releyID, uint8_t hour, uint8_t minute, uint8_t second, uint8_t duration, uint8_t days)
```
Rename `releyID` to `relayID`

Elemeants commented on src/system/storage/storage.h:
```diff
@@ -0,0 +1,67 @@
+} eeprom_map_conf_time_t;
+
+
+String templateId;
+String templateName;
+String authToken;
```
These should be declared on the C file as static variables.

m-mora committed:
implemented feedback:
- move varialbles to cpp an declare them static
- Fix a incorrect calculation for relays space in eeprom
- rename variables

Elemeants reviewed (approved):
LGTM, did you test it?
